Initial Upload of files to GitHub

gyrationtechs opened the pull request.

gyrationtechs committed:
Initial Upload of files to GitHub

sixFingers commented on src/email/email.ts:
```diff
@@ -0,0 +1,190 @@
+import { EmailRecipient, EmailAttachment, EmailOptions, SendEmailResponse } from '../types';
+
+
+export class NewEmail extends BaseClient {
```
I'm curious about this choice of having each API namespace (emails, events, webhooks) extend a base client class, as it results in multiple unneeded instances of Axios. 

All in all, it seems to me that emails/events/webhooks are just collection of methods, so maybe we could structure the whole sdk differently, composing instead of extending: BaseClient becomes the root class, and Emails, Webhook and Events are its 3 public properties.

We'd use it like this:

```
const sendLayer = new SendLayer()

sendLayer.Emails.send(...)
```

or:

```
const { Emails, Webhooks, ... } = new SendLayer()
```

What do you think?

gyrationtechs replied:
@sixFingers This approach you're suggesting makes perfect sense to me. I just thought about it, it doesn't make sense to initialize the API request for each method.


Quick question about implementing review changes. Not entirely sure about the process.
Should I:

1. Implement the changes locally and push to this branch?
2. Edit the files directly from the PR on GitHub?

Could you please share the process with me, or point me to any resources that might be helpful 🙂

sixFingers commented on src/email/email.ts:
```diff
@@ -0,0 +1,190 @@
+      },
+      To: toList,
+      Subject: options.subject,
+      ContentType: options.html ? "HTML" : "Text",
```
Maybe it's worth hiding these "magic" strings behind enums? That would make it more accessible to IDEs.

sixFingers commented on src/email/email.ts:
```diff
@@ -0,0 +1,190 @@
+      To: toList,
+      Subject: options.subject,
+      ContentType: options.html ? "HTML" : "Text",
+      [options.html ? "HTMLContent" : "PlainContent"]: options.html || options.text
```
See my comment about magic strings vs enums.

sixFingers commented on src/email/email.ts:
```diff
@@ -0,0 +1,190 @@
+      [options.html ? "HTMLContent" : "PlainContent"]: options.html || options.text
+    };
+
+    if (options.cc) {
```
We have a bunch of repeated, identical checks for various recipients (cc, bcc, etc). I'm wondering, could we refactor this so that:

- the actual logic is not duplicated
- we don't need to pass a magic string to the `validateRecipient` method just for error output purposes?

Maybe something like:

```
const fields = [ 'cc', 'bcc', ... ]

for (const field of fields) {
    try {
        validation logic
    } catch() {
        throw exception with message that uses field const above
    }
}
```

sixFingers commented on src/email/email.ts:
```diff
@@ -0,0 +1,190 @@
+      for (const currentPath of pathsToTry) {
+        try {
+          if (fs.existsSync(currentPath)) {
+            if (!fs.statSync(currentPath).isFile()) {
```
This nested `if` looks weird, what happens if `currentPath` doesn't exist? Could the double `if` be merged into a single `&&` condition?

gyrationtechs replied:
If `currentPath` doesn't exist, it defaults to the `catch` block part of the logic. The nested `if` logic checks if the `currentPath` is a file. If it's not a file, it avoids making unnecessary `fs.startSync()` calls. AI also suggested using this approach to make the error handling more explicit.

Could you share your thoughts on this and your recommendations also?

sixFingers replied:
I see, I missed the final throw statement a few lines below:
```
throw new SendLayerValidationError(`Attachment file not found: ${filePath}`);
```
Makes sense now :+1:

sixFingers commented on src/events/events.ts:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  async getAll(options: GetEventsOptions = {}): Promise<GetEventsResponse> {
+    const validEventTypes = [
```
See my comment about magic strings vs enums.

sixFingers commented on src/index.ts:
```diff
@@ -0,0 +1,22 @@
+} from './exceptions';
+
+export {
+  NewEmail,
```
I'd rename this to just `Emails` to align it with `Webhooks` and `Events`.

sixFingers commented on src/webhooks/webhooks.ts:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    // Validate event type
+    const eventOptions = ["bounce", "click", "open", "unsubscribe", "complaint", "delivery"];
```
See my comment about magic strings vs enums.

gyrationtechs committed:
Updated SDK architecture to use a centralized root class

gyrationtechs committed:
Updated readme, examples and test cases to match the new SDK structure

gyrationtechs commented:
@sixFingers I have implemented changes following your recommendations. Please review and let me know what you think🙂

sixFingers commented on src/base/client.ts:
```diff
@@ -0,0 +1,81 @@
+export class BaseClient {
+  protected client: AxiosInstance;
+
+  constructor(apiKey: string) {
```
It'd be nice to also allow a second parameter with optional config, so that the consumer can change the underlying Axios instance, like this (pseudo-code-ish):

```
constructor(apiKey: string, config = {})

const defaultConfig = {
  baseURL: 'https://console.sendlayer.com/api/v1/',
  headers: {
    'Authorization': `Bearer ${apiKey}`,
    'Content-Type': 'application/json'
  }
}

this.client = axios.create({...defaultConfig, ...config})
```

sixFingers commented on src/email/email.ts:
```diff
@@ -0,0 +1,209 @@
+      // Check if the path is a URL
+      if (this.isUrl(filePath)) {
+        try {
+          const response = await axios.get(filePath, { responseType: 'arraybuffer', timeout: 30000 });
```
This hard-coded timeout should be configurable, maybe as part of the optional config parameter of the BaseClient constructor. In that case, the optional config parameter could take this form:

```
{
    axios: {...}
    attachmentURLTimeout: 30000,
    ...
}
```

gyrationtechs replied:
@sixFingers Not sure I completely understand this requirement you mentioned. I used AI to try and understand and implement it. Please review the implementation and lemme know what you think. Thanks!

sixFingers commented on package.json:
```diff
@@ -0,0 +1,60 @@
+    },
+    "dependencies": {
+        "axios": "^1.6.7",
+        "dotenv": "^16.4.7"
```
`dotenv` should not be required for production use (consumer should be free to provide env variables however they like), we should move this to devDependencies as it's used only by tests, if I'm not wrong.

sixFingers commented on src/types.ts:
```diff
@@ -0,0 +1,133 @@
+}
+
+export interface EmailOptions {
+    from_email: string;
```
Could we make this uniform with other recipients, so it becomes `from: string | EmailRecipient`?

sixFingers commented on src/types.ts:
```diff
@@ -0,0 +1,133 @@
+    subject: string;
+    html?: string;
+    text?: string;
+    cc?: EmailRecipient | EmailRecipient[];
```
Could we make this uniform with other recipients, so it becomes `string | string[] | EmailRecipient | EmailRecipient[]`?

sixFingers commented on src/types.ts:
```diff
@@ -0,0 +1,133 @@
+    html?: string;
+    text?: string;
+    cc?: EmailRecipient | EmailRecipient[];
+    bcc?: EmailRecipient | EmailRecipient[];
```
Same as previous comment.

sixFingers commented on src/types.ts:
```diff
@@ -0,0 +1,133 @@
+    text?: string;
+    cc?: EmailRecipient | EmailRecipient[];
+    bcc?: EmailRecipient | EmailRecipient[];
+    replyTo?: EmailRecipient | EmailRecipient[];
```
Same as previous comment.

sixFingers commented on src/types.ts:
```diff
@@ -0,0 +1,133 @@
+
+export interface EmailOptions {
+    from_email: string;
+    from_name?: string;
```
This would become obsolete because of my previous commentt.

sixFingers commented:
Hey @gyrationtechs, I've tested and tried various failure scenarios, and overall, it's working great! Here's a few minor things I found:

* In Emails namespace, message errors are sometimes unclear:

```
const response = await sendlayer.Emails.send({})
```

Returns: `Error sending email: Invalid sender email address: undefined
`
```
const response = await sendlayer.Emails.send({
    from_email: 'sender@example.com',
});
```

Returns: `Error sending email: Cannot read properties of undefined (reading 'email')`

```
const response = await sendlayer.Emails.send({
    from_email: 'sender@example.com',
    to: 'recipient@example.com',
});
```

Returns: `Error sending email: Unprocessable Entity`

Maybe we could unify error related to any "address" (name/email pair) so that everything defaults to "invalid [from|to|cc|bcc|...] email address".

* Using this tag format: 

```
const response = await sendlayer.Emails.send({
    ...,
    tags: [
        {field: 'value'}
    ]
});
```

ended up pushing an `Array` tag to Sendlayer.

* Not sure if it's intended, but both this:

```
const simpleResponse = await sendlayer.Emails.send({
    ...,
    tags: [
        'one', 'two'
    ]
});
```

and this:

```
const simpleResponse = await sendlayer.Emails.send({
    ...,
    tags: [
        'one, two'
    ]
});
```

End up pushing 'one' and 'two' tags to Sendlayer.

* This example: 

```
const filteredEvents = await sendlayer.Events.get({
    event: 'sent'
});
```

returns all events for me. I had to use event `delivered`.

* The same goes for Webhooks `event` parameter, running this:

```
// Create a webhook
const webhook = await sendlayer.Webhooks.create({
  url: 'https://sixfingers.me/webhook',
  event: 'open'
});
```

returns an invalid eventType error, I had to use `opened`. Just in case we provide examples, let's make sure they all work.

See also my other comments. Once we've tidied up these last items, this is good to go 🙂 Thanks!

gyrationtechs committed:
Fix: Added validation checks for required payload and tags

gyrationtechs committed:
Changed: eventType to event in payload

gyrationtechs committed:
Fix: Updated supported webhooks event ENUM

gyrationtechs committed:
Updated: READMe, Testcases, and webhooks event validation

gyrationtechs commented:
@sixFingers thanks for the detailed testing and feedback. Here is a brief summary of the changes I made:

The first 4 feedback is related to payload missing required parameters. It triggers the email validation method even if the payload is empty, leading to the `undefined` error you saw. To address it, I added a validation check for required parameters (`from_email`, `to`, `subject`, `text` or `HTML`). Now the SDK won't send an API request if any of the required parameters are missing. 

> ended up pushing an Array tag to Sendlayer.

I added a validation check to ensure tags are array of strings, otherwise it will trigger an error. Regarding the second point, I think it makes sense to retain the logic where `['one, two']` and `['one', 'two']` pushes two tags to the API. Since tags need to be single words.

> This example:
```
const filteredEvents = await sendlayer.Events.get({
    event: 'sent'
});
returns all events for me. I had to use event delivered.
```
This happened because I defined `eventType` instead of `event` as the parameter in the `eventOptions` interface. Leading to the API ignoring the parameter and just rendering all events. It should be working fine now.


> returns an invalid eventType error, I had to use opened. Just in case we provide examples, let's make sure they all work.

This was happening because I had mistakenly defined the same enum for events and webhooks. I've updated it to match the supported events for creating a webhook. In our API docs and the SDK docs we'll release, we've also mentioned the supported events in creating a webhook.

Thanks again. I really appreciate the feedback🙂

gyrationtechs committed:
Updated: EmailOptions interface and email client implementation

gyrationtechs commented:
@sixFingers I've made changes to the SDK based on your suggestion. Please review and lemme know what you think. Thanks!

sixFingers commented:
@gyrationtechs Looking great now! Thanks for your hard work 👍
